src/relay/pass/var_well_formed.cc

```diff
@@ -0,0 +1,55 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file var_well_formed.cc
+ * \brief Function for substituting a concrete type in place of a type ID
+ */
+#include <tvm/relay/pass.h>
+#include <tvm/relay/expr_functor.h>
+#include <unordered_set>
+
+namespace tvm {
+namespace relay {
+
+struct ShadowDetected { };
+
+struct DetectShadow : ExprVisitor {
```

Comment:
Maybe we should also rename this to reflect that fact that is does all well-formedness checks?

Reply:
Could you add a doc-comment describing what checking this performs. It would be good to have a summary of all the checks, so say if we add type well-formedness we could list that there as well. This makes it easier for developers to understand what a well-formed Relay program is too.